### PR TITLE
Add discard=unmap option

### LINKED ISSUE #4781: Enable discard=unmap in vm xml for reclaiming space when using ceph rbd images.
**Is this a BUG REPORT or FEATURE REQUEST?**:

  /kind enhancement

**What happened**:
Ceph-rbd images are thinly provisioned and after mounting in a vm(filesystem or block) they expand accordingly. After deleting files from the images the image size remains the same and thus ends up filling up the storage pool. Fstrim on the mounts also doesn't discard the expanded filesystem.

**What you expected to happen**:

After files are deleted from the images(mounted inside the vm) and on running fstrim the image size should be reclaimed.
**How to reproduce it (as minimally and precisely as possible)**:
1. Add a disk to a vm from ceph rbd storage class
2. Mount it in the vm and create a huge file in it.
3. Check the image size using below command
    ```
    rbd diff <rbd-pool-name>/<volume-name> |
    awk '{ SUM += $2 } END { print SUM/1024/1024/1024 " GB" }' 
    ```
4. Delete the file inside the mount.
5. Repeat step 3.
Detailed steps are in this note: https://hackmd.io/SSUTRWu-QWuj7GDGv27nLw

**Anything else we need to know?**:
The vm xml needs to contain `discard="unmap"` under DiskDriver. Upon testing with the same it works.
Fstrim is able to reclaim the space inside the image.

**Environment**:
- KubeVirt version (use `virtctl version`):
    ```
    Client Version: version.Info{GitVersion:"v0.36.0",         GitCommit:"398cd404dba9ced6c298824a1bf681d387dd3a1c", GitTreeState:"clean", BuildDate:"2020-12-16T15:01:21Z", GoVersion:"go1.13.14", Compiler:"gc", Platform:"linux/amd64"}
    Server Version: version.Info{GitVersion:"v0.34.2", GitCommit:"48956f60864ece98118d7894327caf7513c5b942", GitTreeState:"clean", BuildDate:"2020-12-07T23:39:10Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
    ```
- Kubernetes version (use `kubectl version`):
  Using Openshift CLI
    ```
    oc version
    Client Version: 4.5.6
    Server Version: 4.6.3
    Kubernetes Version: v1.19.0+9f84db3
    ```
- VM or VMI specifications:
    KVM VM running server running Red Hat Enterprise Linux CoreOS 46.82.202010301241-0
- Cloud provider or hardware configuration:
    64 cores
    500 GB memory
    Ceph rbd using OCS as a provisioner
- OS (e.g. from /etc/os-release):
    ```
    NAME="Red Hat Enterprise Linux CoreOS"
    VERSION="46.82.202010301241-0"
    VERSION_ID="4.6"
    OPENSHIFT_VERSION="4.6"
    RHEL_VERSION="8.2"
    PRETTY_NAME="Red Hat Enterprise Linux CoreOS 46.82.202010301241-0 (Ootpa)"
    ID="rhcos"
    ID_LIKE="rhel fedora"
    ANSI_COLOR="0;31"
    CPE_NAME="cpe:/o:redhat:enterprise_linux:8::coreos"
    HOME_URL="https://www.redhat.com/"
    BUG_REPORT_URL="https://bugzilla.redhat.com/"
    REDHAT_BUGZILLA_PRODUCT="OpenShift Container Platform"
    REDHAT_BUGZILLA_PRODUCT_VERSION="4.6"
    REDHAT_SUPPORT_PRODUCT="OpenShift Container Platform"
    REDHAT_SUPPORT_PRODUCT_VERSION="4.6"
    OSTREE_VERSION='46.82.202010301241-0'
    ```
- Kernel (e.g. `uname -a`):
    Linux host 4.18.0-193.28.1.el8_2.x86_64 #1 SMP Fri Oct 16 13:38:49 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by alicefr**
**What this PR does / why we need it**:
This PR adds the option discard=unmap that allows freeing space when the underlying storage supports the trim operation.
However, this option is not suited for all disk types. For example, it should be avoided if the disk is preallocated or thick provisioned. The option is ignored if the PVC or DV has an annotation that contains:
- preallocated disks: `/storage.preallocation: true`
- thick provisioned disks: `/storage.thick-provisioned: true`

For example, CDI set the annotation: `cdi.kubevirt.io/storage.preallocation: true` when the preallocation is configured. Additionally, this annotation allows the user or external tools to set these annotations.

In the other cases, the discard=unmap is set as the default when we use either disk or lun as disk types.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #4781

**Special notes for your reviewer**:

**Release note**:

```release-note
Add discard=unmap option
```

### CODE REVIEW

**Comment by alicefr** [comment]
TODOs:
- adjust the unit test for the xml
- create functional test
- test it with hotplugged volumes

Currently, I have an issue with filesystem-based PVCs, the volumes field in the pvc is not corrected set. I can find the volume name but the PersistentVolumeClaim field is `nil`. I need to find the reason. Otherwise for DVs, and block-based PVC, the option is discarded if the annotation is used.

**Comment by alicefr** [comment]
@maya-r fyi: the current kubevirt repo doesn't have a cdi version that have the preallocation. I simply put the annotation manually on the dv.

**Comment by alicefr** [comment]
A first review is very welcome, especially the new fields added to the `VirtualMachineOptions`, and if you think that the discard=unmap should be added to other disk types. Many thanks in advance!

File: pkg/virt-handler/vm.go [hunk 1]
```diff
@@ -2173,6 +2173,34 @@ func (d *VirtualMachineController) processVmUpdate(origVMI *v1.VirtualMachineIns
 		smbios := d.clusterConfig.GetSMBIOS()
 		period := d.clusterConfig.GetMemBalloonStatsPeriod()
 
+		// Find preallocated or thick-provisioned disks
+		var preallocatedVolumes, thickProvisionedVolumes []string
```
Comment (aglitke) [inline pkg/virt-handler/vm.go:2177]: Is it necessary to keep two separate lists?  What we are really trying to find is a single list of volumes which should have discard passthrough.  Perhaps we should abstract the fact that two different annotations can trigger the condition within a single function.

File: pkg/virt-handler/vm.go [hunk 2]
```diff
@@ -2181,7 +2209,9 @@ func (d *VirtualMachineController) processVmUpdate(origVMI *v1.VirtualMachineIns
 				Sku:          smbios.Sku,
 				Version:      smbios.Version,
 			},
-			MemBalloonStatsPeriod: period,
+			MemBalloonStatsPeriod:   period,
+			PreallocatedVolumes:     preallocatedVolumes,
+			ThickProvisionedVolumes: thickProvisionedVolumes,
```
Comment (aglitke) [inline pkg/virt-handler/vm.go:2214]: In my mind there is no need for two lists.

File: pkg/virt-handler/vm.go (hunk 1 above)
Comment (alicefr) [reply pkg/virt-handler/vm.go]: Yes, you're right it's a bit overkill

**Comment by alicefr** [comment]
I integrated @aglitke 's suggestions, adjusted the unit test, and fixed the issue with the filesystem-based pvc.

**Comment by alicefr** [comment]
Rebased the PR and added @maya-r 's functional tests on top

**Comment by clxu-h3c** [comment]
If want "discard=unmap" to work , the disk's bus must be "scsi" rather than "virtio". But for performance, we usually use virtio.

**Comment by alicefr** [comment]
@cixu-h3c it seems they added the support also for virtio. I actually verified this using virtio by using fstrim inside the guest, and it did free the space.This [blog post](https://chrisirwin.ca/posts/discard-with-kvm-2020/) might be also interesting.

**Comment by alicefr** [comment]
Having a closer look at the test failure, we are missing the Preallocation field. Hence, we just have to wait for a second PR to update the CDI APIs in kubevirt

**Comment by clxu-h3c** [comment]
@alicefr   Thanks，i will see if this feature(discard=unmap)  works for thin-provisioning disk when bus is virtio.

**Comment by clxu-h3c** [comment]
@alicefr   Thanks，i will see if this feature(discard=unmap)  works for thin-provisioning disk when bus is virtio.

**Comment by clxu-h3c** [comment]
I validated this feature based on the following versions of QEMU and libvirt.
qemu version: qemu-kvm-5.0-1.el7.x86_64
libvirt version: libvirt-6.3.0-1.el7.x86_64
I run the fstrim -av command inside the virtual machine,  but doesn't free up the physical space it occupies.

**Releated XML:**
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2' cache='directsync' io='native' discard='unmap'/>
      <source file='/vms/images/clxu-jxckws'/>
      <backingStore/>
      <target dev='vda' bus='virtio'/>
      <hotpluggable state='on'/>
      <alias name='virtio-disk0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x09' function='0x0'/>
    </disk>
      <controller type='scsi' index='1' model='virtio-scsi'>
      <alias name='scsi1'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
    </controller>

**Comment by clxu-h3c** [comment]
I used the local file system to save the disk image and validate this feature based on the following versions of QEMU and libvirt.
qemu version: qemu-kvm-5.0-1.el7.x86_64
libvirt version: libvirt-6.3.0-1.el7.x86_64
I run the fstrim -av command inside the virtual machine,  but doesn't free up the physical space it occupies.
Releated XML info.
driver name='qemu' type='qcow2' cache='directsync' io='native' discard='unmap'
controller type='scsi' index='1' model='virtio-scsi'

**Comment by alicefr** [comment]
@clxu-h3c thanks for reporting! Could you please tell me also which OS are you using for the guest? I'll also have a try

**Comment by alicefr** [comment]
@clxu-h3c  Here how I verified the feature using kubevirt:
1- Create a pvc:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-disk
spec:
  storageClassName: local
  accessModes:
    - ReadWriteOnce
  volumeMode: Filesystem
  resources:
    requests:
      storage: 1Gi
```
2- Create a deployment to create the disk.img and check the size of the image during the test:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: debug
  name: debug
spec:
  replicas: 1
  selector:
    matchLabels:
      app: debug
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: debug
    spec:
      volumes:
      - name: dv
        persistentVolumeClaim:
          claimName: pvc-disk
      containers:
      - image: afrosirh/debug-virt
        name: debug-virt
        command: ["sleep", "100000"]
        securityContext:
          privileged: true
        volumeMounts:
        - mountPath: "/disks"
          name: dv
```
3- Exec in the debug pod and create the image:
```bash
$ cluster-up/kubectl.sh exec -ti debug-6dc86fc6b9-mtgj8 -- bash
$ cd disks
$ qemu-img create disk.img 1G
$ ls -lsh disk.img 
1.0M -rw-r--r--. 1 root root 1.0G Mar  3 16:12 disk.img
```
4- Create a vm that uses this pvc:
```yaml
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm-fedora
  name: vm-fedora
spec:
  running: true
  template:
    metadata:
      labels:
        kubevirt.io/vm: vm-fedora
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: containerdisk
          - disk:
              bus: virtio
            name: cloudinitdisk
          - disk:
              bus: virtio
            name: pvc
        machine:
          type: ""
        resources:
          requests:
            memory: 1024M
      terminationGracePeriodSeconds: 0
      volumes:
      - name: pvc
        persistentVolumeClaim:
          claimName: pvc-disk
      - containerDisk:
          image: registry:5000/kubevirt/fedora-cloud-container-disk-demo:devel
        name: containerdisk
      - cloudInitNoCloud:
          userData: |-
           #cloud-config
           password: fedora
           chpasswd: { expire: False }
        name: cloudinitdisk
```
5- Inside the VM create a filesystem and some large data:
```bash
$ cluster-up/kubectl.sh get vm
NAME        AGE   VOLUME
vm-fedora   66s   
$ cluster-up/kubectl.sh virt console vm-fedora
[root@vm-fedora ~]# lsblk
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
vda    252:0    0    4G  0 disk 
└─vda1 252:1    0    4G  0 part /
vdb    252:16   0  366K  0 disk 
vdc    252:32   0    1G  0 disk 
[root@vm-fedora ~]# mkfs.ext4 /dev/vdc
mke2fs 1.45.5 (07-Jan-2020)
Discarding device blocks: done                            
Creating filesystem with 262144 4k blocks and 65536 inodes
Filesystem UUID: e407ab6c-3730-4252-a370-a82a684de237
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376

Allocating group tables: done                            
Writing inode tables: done                            
Creating journal (8192 blocks): done
Writing superblocks and filesystem accounting information: done
[root@vm-fedora ~]# mount /dev/vdc /mnt/
[  156.610556] EXT4-fs (vdc): mounted filesystem with ordered data mode. Opts: (null)
[root@vm-fedora ~]# dd if=/dev/urandom of=/mnt/largefile
```
6- Check the size in the debug pod:
```bash
$ ls -lhs disk.img 
995M -rw-r--r--. 1 107 107 1.0G Mar  3 16:17 disk.img
```

7- Remove the large file and trigger fstrim:
```bash
[root@vm-fedora ~]# rm /mnt/largefile
rm: remove regular file '/mnt/largefile'? y
[root@vm-fedora ~]# fstrim -v /mnt/
/mnt/: 973.4 MiB (1020678144 bytes) trimmed
```

8- Check in the debug container if the space has been freed:
```bash
$ ls -lhs disk.img 
35M -rw-r--r--. 1 107 107 1.0G Mar  3 16:18 disk.img
```

**Comment by clxu-h3c** [comment]
The Guest OS is CentOS7-x86_64-Minimal-1810. 
I would like to see the dumpxml of disk, could you provide it?

**Comment by alicefr** [comment]
@clxu-h3c I simply used the local storage class. Here, the xml part of the disk:
```xml
    <disk type='file' device='disk' model='virtio-non-transitional'>
      <driver name='qemu' type='raw' cache='none' error_policy='stop' discard='unmap'/>
      <source file='/var/run/kubevirt-private/vmi-disks/pvc/disk.img' index='1'/>
      <backingStore/>
      <target dev='vdc' bus='virtio'/>
      <alias name='ua-pvc'/>
      <address type='pci' domain='0x0000' bus='0x06' slot='0x00' function='0x0'/>
    </disk>
```

**Comment by alicefr** [comment]
@clxu-h3c I think it depends on the guest OS, I tried with [centos](http://mirrors.oit.uci.edu/centos/7.6.1810/isos/x86_64/) and fedora 33. I was able to fstrim in fedora but not in centos. You can check also if it is supported using `lsblk -D`. See https://unix.stackexchange.com/questions/306229/how-do-i-check-trim.

**Comment by maya-r** [comment]
/test pull-kubevirt-e2e-k8s-1.17-rook-ceph
now that the other PR is merged

**Comment by alicefr** [comment]
@maya-r I've to rebase

File: pkg/util/types/pvc.go [hunk 3]
```diff
@@ -76,3 +77,15 @@ func IsSharedPVCFromClient(client kubecli.KubevirtClient, namespace string, clai
 	}
 	return
 }
+
+func IsPreallocated(annotations map[string]string) bool {
+	for a, value := range annotations {
+		if strings.Contains(a, "/storage.preallocation") && value == "true" {
```
Comment (aglitke) [inline pkg/util/types/pvc.go:83]: Should we search for the exact annotation (ie. cdi.kubevirt.io/storage.preallocation)?  Should this value be imported from cdi itself?

**Review by aglitke — COMMENTED** [review]
Looks good.

**Comment by alicefr** [comment]
@mhenriks I had to rebase the PR and it lost the lgtm label. Do you mind approve it again? Many thanks!

**Comment by vladikr** [comment]
Looks good to me. Don't see any major changes from a previously approved version.
Thanks @alicefr 
/lgtm

**Comment by rmohr** [comment]
/lgtm cancel

Has some implicit merge conflicts. @alicefr a rebase should solve it.

**Pull request merged**